src/v/cluster/distributed_kv_stm.h

```diff
@@ -0,0 +1,422 @@
+        // do on log truncations.
+        co_return;
+    }
+
+    std::string_view get_name() const final { return "kv_stm"; }
+    ss::future<iobuf> take_snapshot(model::offset) final { co_return iobuf{}; }
+
+    ss::future<result<model::partition_id>> coordinator_for(Key key) {
+        auto holder = _gate.hold();
+        if (!_is_routing_partition) {
+            co_return errc::invalid_request;
+        }
+        auto units = co_await _lock.hold_read_lock();
+        if (!co_await sync(sync_timeout)) {
+            co_return errc::not_leader;
+        }
+        auto it = _coordinators.find(key);
+        if (it != _coordinators.end()) {
+            const auto& current = it->second;
+            if (
+              current.status
+              == coordinator_assignment_status::migration_in_progress) {
+                // keep the client retrying until migration finishes
+                co_return errc::not_leader;
+            }
+            co_return current.partition;
+        }
+        auto num_partitions = co_await total_partitions();
+        if (num_partitions.has_error()) {
+            co_return num_partitions.error();
+        }
+
+        if (kv_data.size() >= offset_keys_limit) {
+            co_return errc::trackable_offset_limit_exceeded;
+        }
+
+        // new assignment
+        iobuf buf;
+        serde::write(buf, op_type::coordinator_assignment);
+        co_await serde::write_async(buf, key);
+        auto bytes = iobuf_to_bytes(buf);
+        auto result = model::partition_id{
+          murmur2(bytes.c_str(), bytes.length()) % num_partitions.value()};
+
+        auto res = co_await replicate_and_wait(
+          op_type::coordinator_assignment,
+          std::move(buf),
+          coordinator_assignment(
+            key, result, coordinator_assignment_status::assigned));
+
+        if (res != errc::success) {
+            co_return res;
+        }
+        co_return co_await coordinator_for(key);
+    }
+
+    ss::future<result<std::optional<Value>>> value_for(Key key) {
+        auto holder = _gate.hold();
+        auto units = co_await _lock.hold_read_lock();
+        if (!co_await sync(sync_timeout)) {
+            co_return errc::not_leader;
+        }
+        auto it = kv_data.find(key);
+        if (it == kv_data.end()) {
+            co_return std::nullopt;
+        }
+        co_return it->second;
+    }
+
+    ss::future<errc> add_kv(Key key, Value val) {
+        auto holder = _gate.hold();
+        auto units = co_await _lock.hold_read_lock();
+        if (!co_await sync(sync_timeout)) {
+            co_return errc::not_leader;
+        }
+        iobuf record_key;
+        serde::write(record_key, op_type::kv_data);
+        serde::write(record_key, key);
+        co_return co_await replicate_and_wait(
+          op_type::kv_data, std::move(record_key), offset_data(key, val));
+    }
+
+    ss::future<errc> delete_kv(Key key) {
```

Comment:
nit: maybe we do not need `_kv` suffix in all those methods ? The class already suggests its purpouse

Reply:
I like these names.